src/rust.rs

```diff
@@ -965,3 +955,81 @@ pub static SIMPLE_GLOBAL_CLASS: JSClass = JSClass {
     cOps: &SIMPLE_GLOBAL_CLASS_OPS as *const JSClassOps,
     reserved: [0 as *mut _; 3]
 };
+
+#[inline]
+unsafe fn get_object_group(obj: *mut JSObject) -> *mut ObjectGroup {
+    assert!(!obj.is_null());
+    let obj = obj as *mut Object;
+    (*obj).group
+}
+
+#[inline]
+pub unsafe fn get_object_class(obj: *mut JSObject) -> *const JSClass {
+    (*get_object_group(obj)).clasp as *const _
+}
+
+#[inline]
+pub unsafe fn get_object_compartment(obj: *mut JSObject) -> *mut JSCompartment {
+    (*get_object_group(obj)).compartment
+}
+
+#[inline]
+pub unsafe fn get_context_compartment(cx: *mut JSContext) -> *mut JSCompartment {
+    let cx = cx as *mut ContextFriendFields;
+    (*cx).compartment_
+}
+
+#[inline]
+pub fn is_dom_class(class: &JSClass) -> bool {
+    class.flags & JSCLASS_IS_DOMJSCLASS != 0
+}
+
+#[inline]
+pub unsafe fn is_dom_object(obj: *mut JSObject) -> bool {
+    is_dom_class(&*get_object_class(obj))
+}
+
+#[inline]
+pub unsafe fn is_window(obj: *mut JSObject) -> bool {
+    (*get_object_class(obj)).flags & JSCLASS_IS_GLOBAL != 0 && IsWindowSlow(obj)
+}
+
+#[inline]
+pub unsafe fn try_to_outerize(rval: MutableHandleValue) {
+    let obj = rval.to_object();
+    if is_window(obj) {
+        let obj = ToWindowIfWindowProxy(obj);
+        assert!(!obj.is_null());
+        rval.set(ObjectValue(&*obj));
+    }
+}
+
+#[inline]
+pub unsafe fn maybe_wrap_object_value(cx: *mut JSContext, rval: MutableHandleValue) {
+    assert!(rval.is_object());
+    let obj = rval.to_object();
+    if get_object_compartment(obj) == get_context_compartment(cx) {
+        if is_dom_object(obj) {
+            return try_to_outerize(rval);
+        }
+    }
+    assert!(JS_WrapValue(cx, rval));
```

Comment:
We miss a fast path here where the compartment is the same and it's a non-WebIDL object.